napari/layers/_source.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import annotations
+
+from contextlib import contextmanager
+from contextvars import ContextVar
+from typing import TYPE_CHECKING, Optional, Tuple
+
+from pydantic import BaseModel
+
+if TYPE_CHECKING:
+    from magicgui.widgets import FunctionGui
+
+
+class Source(BaseModel):
+    """An object to store the provenance of a layer.
+
+    Parameters
+    ----------
+    path: str, optional
+        filpath/url associated with layer
+    reader_plugin: str, optional
+        name of reader plugin that loaded the file (if applicable)
+    sample: Tuple[str, str], optional
+        Tuple of (sample_plugin, sample_name), if layer was loaded via
+        `viewer.open_sample`.
+    widget: FunctionGui, optional
+        magicgui widget, if the layer was added via a magicgui widget.
+    """
+
+    path: Optional[str] = None
+    reader_plugin: Optional[str] = None
+    sample: Optional[Tuple[str, str]] = None
+    widget: Optional[FunctionGui] = None
+
+    class Config:
+        arbitrary_types_allowed = True
+        frozen = True
+
+
+# layer source context management
+
+_LAYER_SOURCE: ContextVar[dict] = ContextVar('_LAYER_SOURCE', default={})
+
+
+@contextmanager
+def layer_source(**source_kwargs):
+    """Creates context in which all layers will be given `source_kwargs`.
+
```

Comment:
Could you maybe add a little bit more of a note on why we take this approach, just really to help us (me) remember in the future (i.e. why the time is saved with this approach). I'm not so familiar with `ContextVar` so this was a little harder for me to parse

Reply:
added an extended docstring.  The "time saving" here is more about the comparison with stack inspection that was tried briefly (so I didn't mention it in the docstring).  Stack inspection took a post-hoc approach to "go and figure out where the layer came from".   This, by contrast, must be called _before_ the layer is created and simply sets a "temporary state" that the `Layer.__init__` method can query directly (no looping or conditionals requires, just simple attribute access).

Reply:
`ContextVar` is a nice [standard library feature](https://docs.python.org/3/library/contextvars.html) that offers a convenient API specifically designed to track this sort of context-local state.  I think the [PEP (567)](https://www.python.org/dev/peps/pep-0567/#specification) does the best job of explaining it.